File: Projekt1/Java/ReversiAI.java
import java.util.LinkedList;

public class ReversiAI {
	private static int MAX_DEPTH = 16;
	
	private static class Move {
	    int x, y;
	    public Move (int x, int y) {
	    	this.x = x;
	    	this.y = y;
	    }
	    
	    public String toString() {
			return "" + ((char) (x + '1')) + ((char) (y + 'a'));  
		}
	}
	
	public static String minimax(GameState Game, int depth) {
	    MAX_DEPTH = depth;
	    int best = (Game.getCurrentPlayer() == Player.BLACK) ? -100 : 100;
	    Move bestMove = new Move(-1,-1);
	    LinkedList<Move> actions = getAvailableMoves(Game);
	    for(Move m : actions) {
	    	int val = optimize(new GameState(Game), m, -100, 100, 1);
	    	System.out.println(">>" + m.toString() + ": " + val);
	    	if ((val > best && Game.getCurrentPlayer() == Player.BLACK) ||
	    		(val < best && Game.getCurrentPlayer() == Player.WHITE)) {
	 	            best = val;
	 	            bestMove = m;
	 	    }
	    }
	    return bestMove.toString();
	}

	private static int evaluate(GameState Game) {
	    return (Game.blackDisks() - Game.whiteDisks());
	}

	private static int optimize(GameState Game, Move m, int alpha, int beta, int depth) {
	    Game.play(m.toString());
	    if(m.x == -1 || depth > MAX_DEPTH) return evaluate(Game);

	    return (Game.getCurrentPlayer() == Player.BLACK) ?
	        maxValue(Game, alpha, beta, depth):
	        minValue(Game, alpha, beta, depth);
	}

	private static int maxValue(GameState Game, int alpha, int beta, int depth) {
	    int value = -100;
	    LinkedList<Move> actions = getAvailableMoves(Game);
	    for(Move m : actions) {
	    	
	        int best = optimize(new GameState(Game), m, alpha, beta, depth+1);
	        for(int i = 0; i < depth; i++) System.out.print("  ");
	        System.out.println("  " + m.toString() + ": " + best);
	        value = (best > value) ? best : value;
	        if(value >= beta) return value;
	        alpha = (alpha > value) ? alpha : value;
	    }
	    return value;
	}

	private static int minValue(GameState Game, int alpha, int beta, int depth) {
	    int value = 100;
	    LinkedList<Move> actions = getAvailableMoves(Game);
	    for(Move m : actions) {
	        int best = optimize(new GameState(Game), m, alpha, beta, depth+1);
	        for(int i = 0; i < depth; i++) System.out.print("  ");
	        System.out.println("  " + m.toString() + ": " + best);
	        value = (value < best) ? value : best;
	        if(value <= alpha) return value;
	        beta = (beta < value) ? beta : value;
	    }
	    return value;
	}
	
	private static LinkedList<Move> getAvailableMoves(GameState Game) {
	    LinkedList<Move> result = new LinkedList<Move>();
	    for (int i = 0; i < 8; i++) {
	        for (int j = 0; j < 8; j++) {
	            if (Game.viableMove(i, j)) {
	            	result.add(new Move(i, j));
	            }
	        }
	    }
	    return result;
	}

}

File: Projekt1/C/ReversiAI.c
#include "Reversi.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

struct GameNode {
     struct GameNode *parent, *children;
     struct GameState state[8][8];
};

struct Move {
    int x, y;
};

void moveToString(struct Move m, char *output);
void getAvailableMoves(struct GameState *Game, struct Move output[32]);
int evaluate(struct GameState *Game);
int optimize(struct GameState Game, struct Move *m, int alpha, int beta, int depth);
int maxValue(struct GameState *Game, int alpha, int beta, int depth);
int minValue(struct GameState *Game, int alpha, int beta, int depth);

int MAX_DEPTH = 15;

void printdepth(int d) {
    for(int i = 0; i < d; i++) printf("  ");
}

void moveToString(struct Move m, char *output) {
    char str[3];
    str[0] = m.x + '1';
    str[1] = m.y + 'a';
    str[2] = '\0';
    strcpy(output, str);
}

void getAvailableMoves(struct GameState *Game, struct Move output[32]) {
    int count = 0;
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 8; j++) {
            if (viableMove(Game, i, j)) {
                output[count].x = i;
                output[count].y = j;
                count++;
            }
        }
    }
    output[count].x = output[count].y = -1;
}

int evaluate(struct GameState *Game) {
    return (Game->black - Game->white);
}

void minimax(struct GameState *Game, int depth, char *output) {
    MAX_DEPTH = depth;
    struct Move actions[32];
    int bestMoveIndex = 0;
    int best = (Game->currentP == Black) ? -100 : 100;
    getAvailableMoves(Game, actions);
    for (int i = 0; actions[i].x != -1; i++) {
        int val = optimize(*Game, &actions[i], -100, 100, 1);
        // char tmp[3];
        // moveToString(actions[i],tmp);
        // printf(">>%s: %i\n", tmp, val);
        if ((val > best && Game->currentP == Black) ||
            (val < best && Game->currentP == White)) {
            best = val;
            bestMoveIndex = i;
        }
    }

    moveToString(actions[bestMoveIndex], output);
}

int optimize(struct GameState Game, struct Move *m, int alpha, int beta, int depth) {
    Player lastPlayer = Game.currentP;
    char move[3];
    moveToString(*m, move);
    play(&Game, move);
    if(m->x == -1 || depth > MAX_DEPTH) return evaluate(&Game);

    return (Game.currentP == Black) ?
        maxValue(&Game, alpha, beta, depth):
        minValue(&Game, alpha, beta, depth);
}

int maxValue(struct GameState *Game, int alpha, int beta, int depth) {
    int value = -100;
    struct Move actions[32];
    getAvailableMoves(Game, actions);
    for (int i = 0; actions[i].x != -1; i++) {
        int best = optimize(*Game, &actions[i], alpha, beta, depth+1);
        // char tmp[3];
        // moveToString(actions[i],tmp);
        // printdepth(depth);
        // printf("  %c%s: %i \n", (Game->currentP == Black ? 'B' : 'W'), tmp, best);
        value = (best > value) ? best : value;
        if(value >= beta) return value;
        alpha = (alpha > value) ? alpha : value;
    }
    return value;
}

int minValue(struct GameState *Game, int alpha, int beta, int depth) {
    int value = 100;
    struct Move actions[32];
    getAvailableMoves(Game, actions);
    for (int i = 0; actions[i].x != -1; i++) {
        int best = optimize(*Game, &actions[i], alpha, beta, depth+1);
        // char tmp[3];
        // moveToString(actions[i],tmp);
        // printdepth(depth);
        // printf("  %c%s: %i \n", (Game->currentP == Black ? 'B' : 'W'), tmp, best);
        value = (value < best) ? value : best;
        if(value <= alpha) return value;
        beta = (beta < value) ? beta : value;
    }
    return value;
}
